test/jdk/sun/net/www/http/KeepAliveStream/KeepAliveStreamFinalizer.java

```diff
@@ -0,0 +1,293 @@
+        clientHttpCalls(address);
+        if (failureReason != null) {
+            throw new RuntimeException(failureReason);
+        }
+    }
+
+    public static InetSocketAddress startHttpServer() throws Exception {
+        InetAddress localHost = InetAddress.getLoopbackAddress();
+        InetSocketAddress address = new InetSocketAddress(localHost, 0);
+        ServerSocket serverSocket = new ServerSocket();
+        serverSocket.bind(address);
+        server = new Server(serverSocket);
+        server.setDaemon(true);
+        server.start();
+        return (InetSocketAddress) serverSocket.getLocalSocketAddress();
+    }
+
+    public static void doRequest(URL url) throws IOException {
+        HttpsURLConnection c = (HttpsURLConnection)url.openConnection();
+        c.setRequestMethod("POST");
+        c.setDoOutput(true);
+        OutputStreamWriter out = new OutputStreamWriter(c.getOutputStream());
+        out.write("test");
+        out.close();
+        int responseCode = c.getResponseCode();
+        // Fully reading the body causes the HttpsClient to be added to the KeepAliveCache immediately,
+        // which avoids this issue since GC will not finalize the HttpsClient.
+    }
+
+    public static void clientHttpCalls(InetSocketAddress address) throws Exception {
+        try {
+            System.out.println("http server listen on: " + address.getPort());
+            String hostAddr = address.getAddress().getHostAddress();
+            if (hostAddr.indexOf(':') > -1) hostAddr = "[" + hostAddr + "]";
+            String baseURLStr = "https://" + hostAddr + ":" + address.getPort() + "/";
+
+            URL testUrl = new URL(baseURLStr);
+
+            // CheckFinalizeSocketFactory is not a real SSLSocketFactory;
+            // it produces regular non-SSL sockets. Effectively, the request
+            // is made over http.
+            HttpsURLConnection.setDefaultSSLSocketFactory(new CheckFinalizeSocketFactory());
+            // now perform up to 3 requests; with the broken KeepAliveStream finalizer,
+            // the second request usually attempts to use a finalized socket
+            for (int i = 0; i < 3; i++) {
+                System.out.println("Request #" + (i + 1));
+                doRequest(testUrl);
+                System.gc();
+                Thread.sleep(100);
+                if (failureReason != null) break;
+            }
+        } finally {
+            server.srv.close();
+        }
+    }
+
+    static class CheckFinalizeSocket extends SSLSocket {
+        private volatile boolean finalized;
+        public void finalize() throws Throwable {
+            System.out.println("In finalize");
+            super.finalize();
+            finalized = true;
+        }
+
+        @Override
+        public InputStream getInputStream() throws IOException {
+            if (finalized) {
+                System.out.println(failureReason = "getInputStream called after finalize");
```

Comment:
For ease of debugging, perhaps use `System.err.println` instead of `System.out`? That way the `Thread.dumpStack()` which is done on the next line will appear in the same jtreg section as this message.

Reply:
I agree with Jaikiran that since we're not using testng here using `System.err` consistently is probably better.

Reply:
Good point. Replaced `System.out` with `System.err` everywhere.